blog: Add "Godot for AA/AAA game development - What's missing?" (#513)
godotengine/godot-website: collections/_article/whats-missing-in-godot-for-aaa.md

---
title: "Godot for AA/AAA game development - What's missing?"
excerpt: "Godot 4.0 is coming out soon. It includes major improvements all across the board in both features, performance and usability. Still, one of the biggest questions the community has is: How does it compare with mainstream commercial offerings?"
categories: ["news"]
author: Juan Linietsky
image: /storage/blog/covers/whats-missing-in-godot-for-aaa.png
date: 2023-01-16 00:00:00
---

> Godot 4.0 is coming out soon. It includes major improvements all across the board in features, performance and usability. Still, one of the biggest questions the community has is: How does it compare with mainstream commercial offerings?

Author: ```suggestion
Godot 4.0 is coming out soon. It includes major improvements all across the board in features, performance, and usability. Still, one of the biggest questions the community has is: How does it compare with mainstream commercial offerings?
```

The features offered are also a lot more reminiscent of AAA games, such as far more material options and advanced visual effects (including circle DOF, volumetric fog, AMD FSR, etc). Additionally, Godot 4.0 supports advanced global illumination techniques such as lightmapping (including SH lightmapping), Voxel GI (which is fully realtime) and SDFGI (which is a single click, open world GI solution). Screen space GI can be used to enhance the realism even more.

### Physics

After a failed attempt to use Bullet, Godot 4.0 returns to its own physics engine which, despite not being a high end physics engine like PhysX, aims to offer a lot more flexibility and "just works" capabilities to users.

Several features were added to Godot Physics since 3.x, such as softbodies and cylinder shape support, as well as several optimizations to use make use of multiple threads.

The custom physics engine still has a considerable amount of issues remaining but we are working hard to ensure it is in a state deem for shipping when 4.0 reaches stability. It will continue seeing improvements afterwards, during the 4.1+ releases.

> That said, Godot 4.0 introduces the ability to bind custom physics engines in runtime (without recompiling Godot) via GDExtension, so its perfectly possibly for the community to integrate other engines such as PhysX, Jolt or Box2D if need to be.

Reviewer: ```suggestion
That said, Godot 4.0 introduces the ability to bind custom physics engines in runtime (without recompiling Godot) via GDExtension, so it's perfectly possible for the community to integrate other engines such as PhysX, Jolt or Box2D if need to be.
```

* **Mesh streaming**: Models are loaded as low detail (few vertices). As they gradually approach the camera, higher resolution versions are streamed from disk. Models that were not used (displayed) since a while are often just freed and will be loaded again when needed.
* **Animation streaming**: Modern games have long cinematics, which require a lot of animation data. Loading those animations require a lot of memory and loading them takes a lot of time. To avoid this, animations are streamed by generally keeping the first second or two in memory and then new sections needed are loaded on demand as the animation plays. Godot 4.0 supports strong animation compression and animation pages, so most of the work is already done.
* **Audio streaming**: Similar to animation streaming, it requires storing the first second or two of audio and then stream the rest directly from disk.

> Of the avobe, most are relatively straightforward to implement. The most complex is *mesh streaming*, which generally needs to be implemented together with a GPU culling strategy to ensure that very large amounts of models can be drawn at no CPU cost. This is more or less what techniques like *Nanite* do in Unreal, although Godot does not need to implement something that complex to be of use in most cases.

Reviewer: ```suggestion
Of the above, most are relatively straightforward to implement. The most complex is *mesh streaming*, which generally needs to be implemented together with a GPU culling strategy to ensure that very large amounts of models can be drawn at no CPU cost. This is more or less what techniques like *Nanite* do in Unreal, although Godot does not need to implement something that complex to be of use in most cases.
```

> Streaming is the most important feature missing for managing large scenes or open worlds. Without it, Godot users are subject to long loading times (as every texture, model and animation has to load before anything is shown). There is also a risk running out of memory if too many asets are used instead of streaming them.

Reviewer: ```suggestion
Streaming is the most important feature missing for managing large scenes or open worlds. Without it, Godot users are subject to long loading times (as every texture, model and animation has to load before anything is shown). There is also a risk running out of memory if too many assets are used instead of streaming them.
```

As such, this means that low level access to all the rendering server structures need to be exposed via GDExtension. This will allow creating custom renderers or plugging custom code during the rendering steps, which is very useful for custom rendering techniques or post processes.

### Scene Job System

Most of the work done for the Godot 4.0 involved large feature and performance improvements to all the servers (rendering, physics, navigation, etc). Servers are also now multithreaded and optimized. Even asset loading can now be done multithreaded (using multiple threads to load multiple assets).

Still, the scene system (which uses those servers), despite several usability improvements, has not seen significant optimization.

Scenes nodes in Godot are mostly intended to carry complex high level behaviors (such as animation trees, kinematic characters, IK, skeletons, etc) for limited amounts of objects (in the hundreds at much). Currently, no threading happens at all and only a single CPU core is used. This makes it very inefficient.

> This makes it is an ideal target for optimizing with multithreading. There is [an initial proposal](https://github.com/godotengine/godot-proposals/issues/4962) on threaded processing for scene nodes, which should give complex scenes a very significant performance boost.

Reviewer: ```suggestion
This makes it an ideal target for optimizing with multithreading. There is [an initial proposal](https://github.com/godotengine/godot-proposals/issues/4962) on threaded processing for scene nodes, which should give complex scenes a very significant performance boost.
```

### Swarms

Scenes, as mentioned before, are designed complex high level behaviors in the hundreds of instances. Still, sometimes, some games require larger amounts of instances but less complex behaviors instead.

This is needed for some types of game mechanics such as:

> * Projectiles (bullet hell as example)

Reviewer: ```suggestion
* Projectiles (bullet hell for example)
```

### Large team VCS support

Godot text file formats are very friendly to version control. They only write what is needed (no redundant information), keep the ordering of sections and are simple enough to understand changes by just looking at them. Few other technologies work as well in this area.

Despite that, this is far from enough to enable large team collaboration. To enable this, Godot VCS support has to improve in several areas:

* Better integration with the filesystem dock.
* Better real-time refresh of assets if they were modified externally (and checked out).
* Support for permissions and file locking: Git does not support this out of the box, but Git-LFS and perforce do. This feature is essential for large teams, avoiding conflicts and keeping files protected from uninteded modifications (as example, an artist modifying code or a scene they don't own by mistake).

Unless the support for this is solid, using Godot in large teams will remain difficult.

### Commercial asset store

> While for very large studios this is not an area of interest, medium stized studios still rely on significant amount of assets and pre-made functionality. The Asset Library currently existing in Godot only links to projects hosted on GitHub and is unable to be used for commercial assets.

Reviewer: ```suggestion
While for very large studios this is not an area of interest, medium-sized studios still rely on significant amount of assets and pre-made functionality. The Asset Library currently existing in Godot only links to projects hosted on GitHub and is unable to be used for commercial assets.
```

Reviewer 2: `medium-sized` would also be better than `medium sized` here.

Author: Edited the original suggestion and committed from it :)
